### ISSUE TITLE
vmmul instruction subtly broken on ARM64

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#12082]
It's becoming clear that this is the cause of several issues:

https://github.com/hrydgard/ppsspp/issues/11179  Tomb Raider jump issue
https://github.com/hrydgard/ppsspp/issues/9843 Warriors Orochi disappearing officer

What's even weirder is that it appears to affect the Tekken 6 leg trembling, although that needs confirmation:
https://github.com/hrydgard/ppsspp/issues/11179#issuecomment-498507501

Theories about how it might be broken:

  * ARM64 regalloc might have bugs
  * ARM64 emitter may have bugs for high register numbers or something
  * ???

**Comment by hrydgard** [comment]
@unknownbrackets  Should we not set MAP_DIRTY|MAP_NOINIT, rather than just MAP_NOINIT if a load of the destination is avoided in MapReg?

**Comment by unknownbrackets** [comment]
It's fine either way.  `MAP_NOINIT = 2 | MAP_DIRTY,`.

-[Unknown]

**Comment by hrydgard** [comment]
Oh, duh. The things you forget...

**Comment by hrydgard** [comment]
Hm, maybe FMADD gives us too much precision (on ARM64 it's a true MAD instruction, while on ARM32 it's a mul + add), and we should switch to FMUL + FADD... (not that it is 100% accurate either..)

**Comment by unknownbrackets** [comment]
Ah, of course.  Sounds like that helped.

I wonder if vfpu multiply just results in a lower mantissa precision (like only 20 bits) or something.

-[Unknown]

**Comment by unknownbrackets** [comment]
I did also note in #2990 that vtfm and vdot seemed potentially related.  I'm pretty sure the bug always happened the same way on mobile as desktop, but maybe it's all related to vfpu dot precision.

It was tricky in Ridge Racer, since it was hard to tell when it all went to pieces.  But if we know it happens every frame in Tekken 6, maybe we can just log all vfpu multiplications, and replicate them on a PSP.  One way to confirm if the bug is actually related would be to model the error (any slow way) or simply map the values for a save stated frame based on a test PSP program, and see if the result improves.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, these instructions all probably use the same reduced-precision hardware dot product.

Anyway, I'm bumping this to 10.0 because the ARM64 backend now matches the others, at least.

**Comment by unknownbrackets** [comment]
It was pointed out to me that the PSP emulator on the PS3 apparently has a setting for "fast vmmul" that is controlled per game.

-[Unknown]

**Comment by hrydgard** [comment]
Ah, in light of what we now know that makes a lot of sense.

**Comment by unknownbrackets** [comment]
Just some initial tests:

```
[x] FPU multiply mantissa bits:
  Float 3x3: 40610000
  Float 7x7: 407e0100
  Float 11x11: 407fe001
  Float 15x15: 407ffe00
  Float 19x19: 407fffe0
  Float 23x23: 407ffffe
  Float 23 x denormal: 0001fffe
  Float -inf double: ff800000
  Float NaN double: ffffffff

[x] VFPU vector multiply mantissa bits:
  Vector 3x3: 40610000
  Vector 7x7: 407e0100
  Vector 11x11: 407fe001
  Vector 15x15: 407ffe00
  Vector 19x19: 407fffe0
  Vector 23x23: 407ffffe
  Vector 23 x denormal: 00000000
  Vector -inf double: ff800000
  Vector NaN double: ff800001

[x] VFPU vector dot mantissa bits:
  Vector 3x3: 40610000
  Vector 7x7: 407e0100
  Vector 11x11: 407fe001
  Vector 15x15: 407ffe00
  Vector 19x19: 407fffe0
  Vector 23x23: 407ffffe
  Vector 23 x denormal: 00000000
  Vector -inf double: ff800000
  Vector NaN double: 7f800001

[x] VFPU matrix multiply mantissa bits:
  Matrix 3x3: 40610000
  Matrix 7x7: 407e0100
  Matrix 11x11: 407fe001
  Matrix 15x15: 407ffe00
  Matrix 19x19: 407fffe0
  Matrix 23x23: 407ffffe
  Matrix 23 x denormal: 00000000
  Matrix -inf double: ff800000
  Matrix NaN double: 7f800001
```

In this, 3x3 means `0x3FF00000 * 0x3FF00000` and 23x23 means `0x3FFFFFFF * 0x3FFFFFFF`.  It seems VFPU always uses FZ behavior (denormal I tried was `0x0000FFFF`, NAN was `0xFFFFFFFF`.)

PPSSPP currently handles all the same as FPU.  I wonder if any of these edge cases are important for Ridge Racer or Tekken.

-[Unknown]

**Comment by hrydgard** [comment]
Seems likely that all VFPU math is FTZ then? We could probably track this additional FTZ state as a cached state through a block - as VFPU instructions usually come in clumps, we'd only need to change the FTZ mode a few times.

**Comment by LunaMoo** [comment]
0x7f800001 looks familiar, Tekken has an function called "isnanf" where it checks for that value.

**Comment by unknownbrackets** [comment]
Yeah, I suspect it's FTZ.  That said, I suspect it won't affect certain ones - like vmax/min.  I'm already pretty sure they don't muck with NAN.

@LunaMoo that's interesting - if it checks for EXACTLY 0x7f800001, then that behavior might currently be wrong.  Currently, we (often but not always) retain whatever mantissa bits the NAN had.  As noted above, vmax/vmin have this behavior, but it seems like vadd/etc. don't.

-[Unknown]

**Comment by hrydgard** [comment]
pushing further work on this to the next version.
